templates/commands/render_action.go

```diff
@@ -211,87 +210,120 @@ type copyParams struct {
+			logger.Debugf("copyRecursive: walkdirfunc visitor skipped file: %s", relToSrc)
 			return fs.SkipDir
 		}
 
 		if de.IsDir() {
+			// We don't create directories when they're encountered by this WalkDirFunc.
+			// Instead, we create output directories as needed when a file needs to be
+			// placed in that directory.
 			return nil
 		}
 
-		dst := filepath.Join(p.dstRoot, relToSrc)
-
-		// The spec file may specify a file to copy that's deep in a
-		// directory tree, without naming its parent directory. We can't
-		// rely on WalkDir having traversed the parent directory of $path,
-		// so we must create the target directory if it doesn't exist.
+		// The spec file may specify a file to copy that's deep in a directory
+		// tree, (like include "some/deep/subdir/myfile.txt") without including
+		// its parent directory. We can't rely on WalkDir having traversed the
+		// parent directory of $path, so we must create the target directory if
+		// it doesn't exist.
 		inDir := filepath.Dir(dst)
 		if err := mkdirAllChecked(pos, p.rfs, inDir, p.dryRun); err != nil {
 			return err
 		}
-
 		dstInfo, err := p.rfs.Stat(dst)
 		if err == nil {
 			if dstInfo.IsDir() {
 				return model.ErrWithPos(pos, "cannot overwrite a directory with a file of the same name, %q", relToSrc) //nolint:wrapcheck
 			}
-			if !p.overwrite {
-				return model.ErrWithPos(pos, "destination file %s already exists and overwriting was not enabled", relToSrc) //nolint:wrapcheck
+			if !ch.overwrite {
+				return model.ErrWithPos(pos, "destination file %s already exists and overwriting was not enabled with --force-overwrite", relToSrc) //nolint:wrapcheck
 			}
 		} else if !os.IsNotExist(err) {
 			return model.ErrWithPos(pos, "Stat(): %w", err) //nolint:wrapcheck
 		}
-
 		srcInfo, err := p.rfs.Stat(path)
 		if err != nil {
 			return fmt.Errorf("Stat(): %w", err)
 		}
 
-		rf, err := p.rfs.Open(path)
-		if err != nil {
-			return model.ErrWithPos(pos, "Open(): %w", err) //nolint:wrapcheck
-		}
-		defer func() { outErr = errors.Join(outErr, rf.Close()) }()
-
-		if p.dryRun {
-			return nil
-		}
-
 		// The permission bits on the output file are copied from the input file;
 		// this preserves the execute bit on executable files.
-		wf, err := p.rfs.OpenFile(dst, os.O_CREATE|os.O_WRONLY, srcInfo.Mode().Perm())
-		if err != nil {
-			return model.ErrWithPos(pos, "OpenFile(): %w", err) //nolint:wrapcheck
-		}
-		defer func() { outErr = errors.Join(outErr, wf.Close()) }()
+		mode := srcInfo.Mode().Perm()
+		return copyFile(pos, p.rfs, path, dst, mode, p.dryRun)
+	})
+}
 
-		if _, err := io.Copy(wf, rf); err != nil {
-			return fmt.Errorf("Copy(): %w", err)
-		}
+func copyFile(pos *model.ConfigPos, rfs renderFS, src, dst string, mode fs.FileMode, dryRun bool) (outErr error) {
+	readFile, err := rfs.Open(src)
+	if err != nil {
+		return model.ErrWithPos(pos, "Open(): %w", err) //nolint:wrapcheck
+	}
+	defer func() { outErr = errors.Join(outErr, readFile.Close()) }()
```

Comment:
That's neat, never though of altering a return type using a defer

Reply:
Yeah. It doesn't feel super clean to me but afaik it's the only way to return an error from a deferred function.